feat(json report): Add absolute path to json report #1154

mcliment opened the pull request:
In the scope of issue #1154 I added a way to extract the project root from the first `FileLeaf` as I could not get it from anywhere else in the `JsonReport`.

I'll add as well a PR to https://github.com/stryker-mutator/mutation-testing-elements schema if you agree on this PR.

Mobrockers commented:
Hi!

Every FolderComposite should have it's FullPath property set. Why could you not use this? Is something wrong with the way we set the FullPath?

mcliment commented:
Hi @Mobrockers, thanks for the quick response.

I see that actually `FolderComposite` also has a relative path to project set so I could use that instead of the `FileLeaf`, so probably is a safe bet. I didn't saw it in the beginning. Thanks for the tip.

Mobrockers commented:
> Hi @Mobrockers, thanks for the quick response.
> 
> I see that actually `FolderComposite` also has a relative path to project set so I could use that instead of the `FileLeaf`, so probably is a safe bet. I didn't saw it in the beginning. Thanks for the tip.

You can use the FullPath instead of using the RelativePath to build the fullpath, but yes.

mcliment commented:
I see a problem here:

![imatge](https://user-images.githubusercontent.com/63848/96892799-4a541980-148a-11eb-9f58-ec86ecd6881b.png)

This is supposed to be the root folder of the project and the first file contained there is:

![imatge](https://user-images.githubusercontent.com/63848/96892653-21cc1f80-148a-11eb-95ec-654e37d82c06.png)

As you can see the `FullPath` of the first folder is not right, in this case should be `C:\Projects\ToyProject\ToyProject\ToyProjectLib1` instead of just `C:\Projects\ToyProject\ToyProject`.

The .csproj of the project is actually in `C:\Projects\ToyProject\ToyProject\ToyProjectLib1`.

Mobrockers commented:
Right, I forgot about that. I think it's valid to just always take the first FolderComposite in the IReadonlyInputComponent. @richardwerkman ?

richardwerkman commented:
We add a FolderComposite on top of all files and folders to wrap other projects in. When run on a single project it looks a bit odd indeed. So when run on a single project you could always take to first FolderComposite from the root and that will be the folder where your .csproj is located.

mcliment commented:
According to your comments and what I understand `RelativePath` and `RelativePathToProjectFile` should be, I made some changes to `InputFileResolver` to match those. Actually, subfolders were lost due to using `sub` instead of `folder` when building the `FullPath` property.

What do you think?

Mobrockers commented on src/Stryker.Core/Stryker.Core/Initialisation/InputFileResolver.cs:
```diff
@@ -143,8 +143,8 @@ private FolderComposite FindProjectFilesUsingBuildalyzer(ProjectAnalyzerResult a
                 Name = string.Empty,
-                FullPath = projectRoot,
-                RelativePath = string.Empty,
+                FullPath = projectUnderTestDir,
```
We need a top level 'container' FolderComposite in case there are multiple projects, so we can't just take the projectUnderTestDir as the root folder. It needs to be one folder level higher.

mcliment replied:
I did some changes, I think it does more sense now. 

But then, I don't see the meaning in calling the property `ProjectRoot` as it's a bit misleading as paths are relative to the parent that holds all the projects in the case you have many.

Mobrockers replied:
You are correct, it's probably a bad name. How about solutionRoot?

mcliment replied:
Done!

Mobrockers replied:
Hi sorry, I thought you were talking about the naming of the projectRoot variable in InputFileResolver.cs.


In the context of the mutation report schema I don't think it makes sense to call it SolutionRoot instead of ProjectRoot as dotnet has solutions and solution files but other languages and frameworks do not have this concept. The mutation report schema is generic, and in other places where we deal with the schema we call the 'combination' of 'modules' a project. 

Perhaps we should have the discussion about what we call it in mutation-testing-metrics instead of here since we are not the only consumers of the json schema.

Mobrockers commented:
Looks good to me. When it's been added to the schema and a new schema version has been released we can merge this into stryker.net as well.

mcliment commented:
Hi @Mobrockers, let's recap a bit on the original goal of the PR and the current status, so I can find out a simple yet effective solution to the current problem.

I have a Toy project with this structure:
```
ToyProjectLib1
\- Folder1
   \- Folder2
      \- Multiplier.cs
\- Adder.cs
```

Using the `master` branch code and setting the working directory on the test projects of my Toy project (that tests `Adder.cs` as well as `Multiplier.cs` I get this structure in the `mutationReport` parameter of the `JsonReport` constructor:

```
Type: FolderComposite
FullPath: null,
Name: null,
RelativePath: null,
RelativePathToProjectFile: null
Children: 3
\- Type: FolderComposite
   FullPath: "C:\Projects\ToyProject\ToyProject",
   Name: "",
   RelativePath: "",
   RelativePathToProjectFile: "."
   Children: 1
   \- Type: FileLeaf
      FullPath: "C:\Projects\ToyProject\ToyProject\ToyProjectLib1\Adder.cs",
      Name: "Adder.cs",
      RelativePath: "Adder.cs",
      RelativePathToProjectFile: "Adder.cs"
\- Type: FolderComposite
   FullPath: "C:\Projects\ToyProject\ToyProject\ToyProjectLib1\Folder1",
   Name: "Folder1",
   RelativePath: "ToyProjectLib1\\Folder1",
   RelativePathToProjectFile: "Folder1"
   Children: 1
   \- Type: FolderComposite
      FullPath: "C:\Projects\ToyProject\ToyProject\ToyProjectLib1\Folder2", // Wrong!!!
      Name: "Folder2",
      RelativePath: "ToyProjectLib1\Folder2",
      RelativePathToProjectFile: "Folder2"
      Children: 1
      \- Type: FileLeaf
         FullPath: "C:\Projects\ToyProject\ToyProject\ToyProjectLib1\Folder1\Folder2\Multiplier.cs",
         Name: "Multiplier.cs",
         RelativePath: "ToyProjectLib1\Folder2\Multiplier.cs", // Wrong!!!
         RelativePathToProjectFile: "Folder1\Folder2\Multiplier.cs"
\- Type: FolderComposite
   FullPath: "C:\Projects\ToyProject\ToyProject\ToyProjectLib1\obj",
   Name: "obj",
   RelativePath: "ToyProjectLib1\obj",
   RelativePathToProjectFile: "obj"
   Children: ...
```

And the computed `Files` for the report are like this:

```
[
    { "Key": "Adder.cs", Value: object },
    { "Key": "ToyProjectLib1\Folder2\Multiplier.cs", Value: object }
]
```

With my changes to the `InputFileResolver` included in this PR, the results are like this:

```
Type: FolderComposite
FullPath: null,
Name: null,
RelativePath: null,
RelativePathToProjectFile: null
Children: 2
\- Type: FolderComposite
   FullPath: "C:\Projects\ToyProject\ToyProject",
   Name: "",
   RelativePath: "",
   RelativePathToProjectFile: null
   Children: 0
\- Type: FolderComposite
   FullPath: "C:\Projects\ToyProject\ToyProject\ToyProjectLib1",
   Name: "ToyProjectLib1",
   RelativePath: "ToyProjectLib1",
   RelativePathToProjectFile: "."
   Children: 3
   \- Type: FileLeaf
      FullPath: "C:\Projects\ToyProject\ToyProject\ToyProjectLib1\Adder.cs",
      Name: "Adder.cs",
      RelativePath: "ToyProjectLib1\Adder.cs",
      RelativePathToProjectFile: "Adder.cs"
   \- Type: FolderComposite
      FullPath: "C:\Projects\ToyProject\ToyProject\ToyProjectLib1\Folder1",
      Name: "Folder1",
      RelativePath: "ToyProjectLib1\Folder1",
      RelativePathToProjectFile: "Folder1"
      Children: 1
      \- Type: FolderComposite
         FullPath: "C:\Projects\ToyProject\ToyProject\ToyProjectLib1\Folder1\Folder2",
         Name: "Folder2",
         RelativePath: "ToyProjectLib1\Folder1\Folder2",
         RelativePathToProjectFile: "Folder1\Folder2"
         Children: 1
         \- Type: FileLeaf
            FullPath: "C:\Projects\ToyProject\ToyProject\ToyProjectLib1\Folder1\Folder2\Multiplier.cs",
            Name: "Multiplier.cs",
            RelativePath: "ToyProjectLib1\Folder1\Folder2\Multiplier.cs",
            RelativePathToProjectFile: "Folder1\Folder2\Multiplier.cs"
   \- Type: FolderComposite
      FullPath: "C:\Projects\ToyProject\ToyProject\ToyProjectLib1\obj",
      Name: "obj",
      RelativePath: "ToyProjectLib1\obj",
      RelativePathToProjectFile: "obj"
      Children: ...
```

And the computed `Files` for the report are like this:

```
[
    { "Key": "ToyProjectLib1\Adder.cs", Value: object },
    { "Key": "ToyProjectLib1\Folder1\Folder2\Multiplier.cs", Value: object }
]
```

So in the end, the value we need to add to the Json report `ProjectRoot` property should be `C:\Projects\ToyProject\ToyProject\ToyProjectLib1` in my understanding should be the `FullPath` of the `FolderComposite` from root that has `RelativeToProjectFile == "."' . Maybe I'm wrong.

As well, I don't really understand the meaning of the first `FolderComposite` that points to `C:\Projects\ToyProject\ToyProject` and has 0 children.

Sorry for the long explanation but I think it's clearer this way to explain the problems I have with the original composite structure.

mcliment commented:
I changed a bit the `GenerateFileReportComponents` in the end so the path to the file is the newly introduced `ProjectRoot` + the key of the Files collection. I think makes sense that way.

As well, I added a PR in `mutation-testing-elements`: https://github.com/stryker-mutator/mutation-testing-elements/pull/698

Thanks for your help.

Mobrockers commented on src/Stryker.Core/Stryker.Core/Reporters/JsonReporter/JsonReport.cs:
```diff
@@ -95,12 +98,21 @@ public string ToJsonHtmlSafe()
         private IDictionary<string, JsonReportFileComponent> GenerateFileReportComponents(ReadOnlyFileLeaf fileComponent)
         {
-            return new Dictionary<string, JsonReportFileComponent> { { fileComponent.RelativePath, new JsonReportFileComponent(fileComponent) } };
+            return new Dictionary<string, JsonReportFileComponent> { { fileComponent.RelativePathToProjectFile, new JsonReportFileComponent(fileComponent) } };
```
Thank you for that! Nice bug fix :)

Mobrockers commented on src/Stryker.Core/Stryker.Core/Initialisation/CsharpProjectComponentsBuilder.cs:
```diff
@@ -105,7 +105,7 @@ private FolderComposite FindProjectFilesUsingBuildalyzer(ProjectAnalyzerResult a
                 }
 
-                var relativePath = Path.GetRelativePath(projectUnderTestDir, sourceFile);
+                var relativePath = Path.GetRelativePath(projectRoot, sourceFile);
```
I don't think this is correct. the projectRoot is the 'solution folder'. Project files are not relative to the solution folder, the are relative to the project folder. Which is the projectUnderTestDir.

Mobrockers replied:
I will take a look at our project structure to see if I can make some sense of it. It is perhaps too deep in stryker for a contributor to dig in to with the amount of available documentation (none 😅)

mcliment replied:
I see things are a bit broken with some inconsistencies here and there because to my understanding, `RelativePath` means relative to root folder and `RelativePathToProjectFile` means relative to project folder, so they can't be the same unless the project is on the root folder.

For example, in line 116, this is how `RelativePath` is set: `RelativePath = _fileSystem.Path.Combine(folderComposite.RelativePath, fileName),` which leads to a wrong path.

Maybe I don't understand exactly what we're trying to achieve here.

Mobrockers replied:
A contributor added the RelativePathToProjectFile at some point and we failed to question why he did that. At this point I'm not entirely sure what we're trying to achieve here either so I'm gonna have to look it over with a colleague and decide on how we want the design to be. Then fix up the whole thing and document what's what and why so we don't mess it up again. It might take us some time to get to that. Apologies.

Mobrockers commented:
TODO:

Fix cleartext-tree reporter after project components refactor
Check that project root still works

richardwerkman commented:
/azp run

richardwerkman commented:
@Mobrockers I think I've cleared all issues with this PR. Could you check it again?

Mobrockers commented on src/Stryker.Core/Stryker.Core/StrykerRunner.cs:
```diff
@@ -134,5 +134,28 @@ private void AnalyseCoverage(StrykerOptions options)
+        /// <param name="projectComponents">A list of all project root components</param>
+        /// <param name="options">The current stryker options</param>
+        /// <returns>The root folder component</returns>
+        private CsharpFolderComposite AddRootFolderIfMultiProject(IEnumerable<IProjectComponent> projectComponents, StrykerOptions options)
```
This should return an IFolderComposite or IProjectComponent

Mobrockers commented on src/Stryker.Core/Stryker.Core/StrykerRunner.cs:
```diff
@@ -134,5 +134,28 @@ private void AnalyseCoverage(StrykerOptions options)
+        /// <returns>The root folder component</returns>
+        private CsharpFolderComposite AddRootFolderIfMultiProject(IEnumerable<IProjectComponent> projectComponents, StrykerOptions options)
+        {
+            var rootComponent = new CsharpFolderComposite
```
It would be better to have a new type for this (RootFolderComposite?) because CsharpFolderComposite is specific to csharp, which is not needed at this stage in the process.

richardwerkman replied:
Why do we even have a CsharpFolder? A folder should not care what type of project it is in right? It can just contain files. The files are indeed language specific.

Mobrockers replied:
Because we use the syntaxtrees on folder level. For example, when compiling we take all syntaxtrees from the root component. The syntaxtree type is csharp specific, so the folder must be as well.

Mobrockers replied:
We also use the folder composite to add compilation only syntax trees to the project (for example the stryker helpers)

Mobrockers commented:
@mcliment Thank you for starting the work on this PR and for showing us all the mess we had to clean up 😅
Sorry it took so long

mcliment commented:
@Mobrockers @richardwerkman Thanks to you guys for maintaining this project!
